File: ios/Sources/WalkySim/ConvexHull.swift
import Foundation

/// Andrew's monotone chain convex hull, O(n log n). Ports `src/sim/convexHull.ts`.
///
/// Returns a *strictly* convex hull: collinear points along an edge are
/// dropped, which is what makes the offset step in `expandPolygon` well defined.
///
/// JS `Array.sort` is stable and Swift's `sorted` is not, which does not matter
/// here: the comparator is a total order on distinct points, and exact
/// duplicates are dropped immediately afterwards, so no two elements that
/// compare equal survive to be distinguished.
public func monotoneChainHull(_ points: [Point]) -> [Point] {
  if points.isEmpty { return [] }

  let pts = points.sorted { a, b in a.x != b.x ? a.x < b.x : a.y < b.y }
  var uniq: [Point] = []
  uniq.reserveCapacity(pts.count)
  for p in pts {
    if let last = uniq.last, last.x == p.x, last.y == p.y { continue }
    uniq.append(p)
  }
  let n = uniq.count
  if n < 3 { return uniq.map { Point($0.x, $0.y) } }

  func build(_ src: [Point]) -> [Point] {
    var chain: [Point] = []
    chain.reserveCapacity(src.count)
    for p in src {
      // <= 0 drops collinear points, giving a strictly convex hull.
      while chain.count >= 2 && orient(chain[chain.count - 2], chain[chain.count - 1], p) <= 0 {
        chain.removeLast()
      }
      chain.append(p)
    }
    chain.removeLast()   // the last point belongs to the other chain
    return chain
  }

  let hull = build(uniq) + build(uniq.reversed())

  // All input collinear: both chains collapse to the two extreme points.
  if hull.count < 3 {
    return [Point(uniq[0].x, uniq[0].y), Point(uniq[n - 1].x, uniq[n - 1].y)]
  }
  return hull.map { Point($0.x, $0.y) }
}

File: web/src/sim/convexHull.ts
import { orient, type Point } from './geometry';

/**
 * Andrew's monotone chain convex hull, O(n log n).
 *
 * Replaces the original's ConvexHullGenerator.createQuickHull (adapted from a 2007
 * web posting), which is O(n^2) in the worst case and mishandles collinear input.
 * Monotone chain sorts once and builds two chains with an exact integer orientation
 * test, so collinear runs, duplicate points and degenerate inputs are all decided
 * without an epsilon.
 *
 * Returns a strictly convex hull: collinear points along an edge are dropped, which
 * is what makes the offset step in expandPolygon well defined.
 */
export function monotoneChainHull(points: readonly Point[]): Point[] {
  if (points.length === 0) return [];

  // Sort by x, then y, and drop exact duplicates.
  const pts = [...points].sort((a, b) => (a[0] - b[0]) || (a[1] - b[1]));
  const uniq: Point[] = [];
  for (const p of pts) {
    const last = uniq[uniq.length - 1];
    if (!last || last[0] !== p[0] || last[1] !== p[1]) uniq.push(p);
  }
  const n = uniq.length;
  if (n < 3) return uniq.map((p) => [p[0], p[1]] as Point);

  const build = (src: Point[]): Point[] => {
    const chain: Point[] = [];
    for (const p of src) {
      // <= 0 drops collinear points, giving a strictly convex hull.
      while (chain.length >= 2 && orient(chain[chain.length - 2], chain[chain.length - 1], p) <= 0) {
        chain.pop();
      }
      chain.push(p);
    }
    chain.pop(); // last point belongs to the other chain
    return chain;
  };

  const lower = build(uniq);
  const upper = build([...uniq].reverse());
  const hull = lower.concat(upper);

  // All input collinear: both chains collapse to the two extreme points.
  if (hull.length < 3) {
    return [uniq[0], uniq[n - 1]].map((p) => [p[0], p[1]] as Point);
  }
  return hull.map((p) => [p[0], p[1]] as Point);
}
